src/config/ConfigBuilder.ts

```diff
@@ -91,15 +95,85 @@ export class ConfigBuilder {
             } else if (key === 'customLogger') {
                 this.handleLogger(value);
             } else if (key === 'customCredentials') {
-                this.effectiveConfig.customCredentials = jsonObject;
+                this.effectiveConfig.customCredentials = value;
             } else if (key === 'backupAckToClientEnabled') {
                 this.effectiveConfig.backupAckToClientEnabled = tryGetBoolean(value);
+            } else if (key === 'security') {
+                this.handleSecurity(value);
             } else {
                 throw new RangeError(`Unexpected config key '${key}' is passed to the Hazelcast Client`);
             }
         }
     }
 
+    private static validateSecurityConfiguration(jsonObject: any): void {
+        if ('security' in jsonObject && 'customCredentials' in jsonObject) {
+            throw new RangeError('Ambiguous security configuration is found. ' +
+                'Use one of \'security\' or \'customCredentials\' elements, not both.')
+        }
+    }
+
+    private handleSecurity(jsonObject: any): void {
```

Comment:
So if we have the `security` thing, we must have `usernamePassword`, `token` or `custom`... and what is `custom` exactly, is it the thing that enables us to inject any kind of credentials factory?

Reply:
Yes, exactly, if you have something in the `security` element, it must be one of those three.

`custom` simply means any kind of object: portable, identified etc. where we serialize it as it is and send it to the server. It is similar to credentials factory in that regard